app/src/org/commcare/android/tasks/ManageKeyRecordTask.java

```diff
@@ -118,100 +132,113 @@ protected void deliverError(R receiver, Exception e) {
+        }
+
+        /* Should only try to look for new records if ALL of the following are true:
+         * a) We're in normal password login mode (otherwise, should only be try matching to an existing record on the device)
+         * b) We didn't find a matching record that is valid
+         * c) There is a keyServerUrl to make the http callout to */
+        calloutNeeded = (loginMode == LoginActivity.LoginMode.PASSWORD)
+                && (!hasRecord || valid == null)
+                && keyServerUrl != null;
         
-        if(calloutNeeded) {
-            Logger.log(AndroidLogger.TYPE_USER, "Performing key record callout." + (calloutRequired ? " Success is required for login" : ""));
+        if (calloutNeeded) {
+            calloutSuccessRequired = !hasRecord;
+            Logger.log(AndroidLogger.TYPE_USER, "Performing key record callout." + (calloutSuccessRequired ? " Success is required for login" : ""));
             this.publishProgress(Localization.get("key.manage.callout"));
         }
         
         return null;
     }
 
-    
-    //Functionality that doesn't necessarily "belong" to this workflow:
-
     private void cleanupUserKeyRecords() {
         UserKeyRecord currentlyValid = null;
         //For all "new" entries: If there's another sandbox record (regardless of user)
         //which shares the sandbox ID, we can set the status of the new record to be
         //the same as the old record.
         
-        //TODO: We dont' need to read these records, we can read the metadata straight.
         SqlStorage<UserKeyRecord> storage = app.getStorage(UserKeyRecord.class);
-        for(UserKeyRecord record : storage) {
-            if(record.getType() == UserKeyRecord.TYPE_NORMAL) {
-                if(record.getUsername().equals(username) && record.isCurrentlyValid() && record.isPasswordValid(password)) {
-                    if(currentlyValid == null) {
-                        currentlyValid = record;
-                    } else {
-                        Logger.log(AndroidLogger.TYPE_ERROR_ASSERTION, "User " + username + " has more than one currently valid key record!!");
-                    }
-                }
-            } else if(record.getType() == UserKeyRecord.TYPE_NEW) {
-                //See if we have another sandbox with this ID that is fully initialized.
-                if(storage.getIDsForValues(new String[] {UserKeyRecord.META_SANDBOX_ID, UserKeyRecord.META_KEY_STATUS}, new Object[] {record.getUuid(), UserKeyRecord.TYPE_NORMAL}).size() > 0) {
-                    
-                    Logger.log(AndroidLogger.TYPE_MAINTENANCE, "Marking new sandbox " + record.getUuid() + " as initialized, since it's already in use on this device");
-                    //If so, this sandbox _has_ to have already been initialized, and we should treat it as such.
-                    record.setType(UserKeyRecord.TYPE_NORMAL);
-                    storage.write(record);
+
+        for (UserKeyRecord normalRecord :

```

Comment:
just refactored this to use getRecordsForValue() for each TYPE, rather than using if statements to check for each type